File: fused_render/server/routers/selffix.py
"""Self-fix — the HTTP surface over `fused_render/selffix.py` (SPEC §43).

Three routes, and the split between them is what each costs:

  POST /api/selffix/start   start a Claude session on this installation
  GET  /api/selffix         everything the modified-version panel shows
  POST /api/selffix/clear   forget the modification (the user's own override)

One session at a time (`_claim_active`): they all edit the same tree, so a
second one is not concurrency, it is a conflict.

The BADGE itself is not here: whether the chip is amber rides on /api/config's
`modified_install` field, because the shell already polls that endpoint on every
route and a second poll for one boolean would be a second poll for one boolean.
What this endpoint answers is the panel's contents — the report list, the
reinstall instructions — which cost a directory walk and (on a mac) a `brew`
subprocess, and are read once, when the user opens the panel.

Reads are unguarded and writes carry the X-Fused header, exactly as everywhere
else.
"""
from __future__ import annotations

import logging
import os
import threading

from fastapi import APIRouter, Body, Header

from fused_render import claude_spawn, selffix
from fused_render.server import templates as _server_templates
from fused_render.server.common import _error, _require_fused

logger = logging.getLogger(__name__)

router = APIRouter()

# How often the watcher re-hashes the installation while a session runs, in
# poll ticks (claude_spawn polls every 2s, so ~16s).
#
# There IS a reason to check before the turn ends, and it is not impatience: the
# user is watching this session in the sidebar, and the badge appearing while
# they watch is what tells them the app noticed. The digest is a walk of the
# package (~460 files), so it is not something to do on every tick — but once
# every few seconds of a session someone is sitting in front of is nothing.
_CHECK_EVERY_TICKS = 8

# ONE fix session at a time, because they all edit the same tree. Two agents
# rewriting one installation concurrently is not a slow path, it is a conflict:
# each reads a file the other is midway through changing, and the report each
# writes describes a state that never existed. A user with two failed rows
# clicking Fix on both is the ordinary way to get there.
#
# THE QUESTION IS ASKED, NOT REMEMBERED. This used to be a module-global claim
# with a token and a TTL, and it was wrong in the one way that mattered: the
# claude process is DETACHED and outlives this server, while the claim lived in
# memory and did not. A restart cleared the guard with an agent still editing —
# and the restart most likely to happen is one the fix session CAUSES, since it
# edits .py files under a dev server watching them. Persisting a lease instead
# would mean a lease file, a TTL for it, recovery at startup and a way to tell a
# stale lease from a live one: four pieces of state to keep in step with one
# fact that is already on disk.
#
# `agent._live_run(dir)` is that fact. Every run writes `RUNS/<id>/meta.json`
# with its target and a `pid` file, so "is a Claude session working in this
# directory?" is a directory scan plus a liveness check — durable across
# restarts by construction, with nothing to expire, release or recover. It also
# answers the question we actually care about rather than a proxy for it: a chat
# the user opened on the install folder BY HAND is another agent in the same
# tree, and the old claim could not see it.
#
# The lock below is not that state. It is held only for the check-and-spawn of a
# single request, so two simultaneous clicks cannot both look, both see nothing,
# and both spawn; it is released in a `finally` a second later and holds nothing
# about the session that is now running.
_spawn_lock = threading.Lock()


# Re-bound as module-level names (the apps router's convention) so a test can
# swap the spawn without reaching into another module.
_spawn_helper = claude_spawn.spawn_helper
_load_agent = claude_spawn.load_agent
_record_session_when_ready = claude_spawn.record_session_when_ready


def _claude_cli_path():
    """The CLI this machine would run, or None — re-bound for the tests, like
    the spawn above. Resolved by the same function /api/config reports from, so
    the button's wording and this refusal cannot disagree."""
    from fused_render.server import ai

    return ai.claude_cli_path()


def _live_session_in(root: str) -> str:
    """The id of a Claude run still going in `root`, or "" if there is none.

    TWO LOOKUPS, and they are not two answers to be reconciled — either one
    saying "busy" is enough, because each covers a case the other cannot.

      the run we RECORDED   `selffix.active_run()` names the fix started here,
                            however long ago. Checked by pid, so the record is
                            a pointer and never a lease.
      the runs DIRECTORY    `agent._live_run(root)` finds anything else live in
                            this tree — including a chat the user opened on the
                            install folder by hand, which nothing we record
                            could know about.

    The recorded half exists because the scan is BOUNDED: `_live_run` reads the
    newest `_LIVE_SCAN_LIMIT` runs on the machine before filtering by target,
    which is right for a chat turn ("a turn does not outlive 60 later ones") and
    wrong for this one. A fix session is the long-running case by construction,
    and a machine running scheduled tasks can start 60 runs beside it — at which
    point the scan stops seeing the very session it is meant to be guarding.

    Never raises: a lookup that cannot answer must not be the reason a fix
    cannot start. Failing OPEN is the right direction here and not a coin toss —
    everything this can fail on (the agent not loading, the runs dir being
    unreadable) fails the spawn a moment later too, with a message that says
    what actually went wrong instead of "already running".
    """
    try:
        agent = _load_agent()
    except Exception:  # noqa: BLE001 — see the docstring
        logger.debug("could not load the agent to check for a live session",
                     exc_info=True)
        return ""

    recorded = selffix.active_run()
    if recorded:
        try:
            if agent._alive(os.path.join(agent.RUNS, recorded)):
                return recorded
        except Exception:  # noqa: BLE001
            logger.debug("liveness check failed for run %s", recorded, exc_info=True)

    try:
        return str(agent._live_run(root).get("run_id") or "")
    except Exception:  # noqa: BLE001
        logger.debug("live-run scan failed for %s", root, exc_info=True)
        return ""


def _watch_fix(run_id: str, incident: str, report: str, title: str,
               before: str) -> None:
    """Follow the fix session and stamp the installation if it changed.

    Two jobs the session cannot do for itself. The FIRST is the stamp — see
    `selffix.settle` for why the app decides that rather than the model, and why
    it is measured against `before` (the tree as THIS session found it) rather
    than against the release. The SECOND is the one every detached session in
    this codebase needs: nobody polls a run started from an HTTP request, and a
    run nobody polls never reaches its sidecar, so the conversation would not be
    listed when the user later opens that folder's chat.

    Every failure is swallowed. A fix that landed but whose badge did not is a
    bad outcome; a background thread that raises into the server is a worse one.
    """
    state = {"ticks": 0, "session_id": ""}

    def stamp() -> None:
        try:
            selffix.settle(before=before, run_id=run_id,
                           session_id=state["session_id"],
                           report=report, incident=incident, title=title)
        except Exception:  # noqa: BLE001 — bookkeeping, never fatal
            logger.debug("self-fix stamp failed", exc_info=True)

    def on_tick(data: dict) -> bool:
        session_id = str(data.get("session_id") or "")
        if session_id:
            state["session_id"] = session_id
        state["ticks"] += 1
        # The `done` tick is checked first and unconditionally: it is the only
        # one guaranteed to happen, and a session that edits one file in its
        # last thirty seconds must still be caught.
        if data.get("done") or state["ticks"] % _CHECK_EVERY_TICKS == 0:
            stamp()
        return True

    try:
        _record_session_when_ready(_load_agent(), run_id, on_tick=on_tick)
    except Exception:  # noqa: BLE001
        logger.debug("self-fix watcher failed", exc_info=True)
    finally:
        # One last look regardless of how the poll ended — it gives up after an
        # hour (claude_spawn._RECORD_POLL_TICKS), and a session that ran long is
        # exactly the one most likely to have changed something.
        #
        # This thread no longer releases anything: the guard is a question asked
        # of the runs directory, so a watcher that dies takes nothing with it and
        # a session it stopped following still excludes a second one for exactly
        # as long as its process is alive.
        stamp()


@router.post("/api/selffix/start")
def api_selffix_start(body: dict = Body(default={}),
                      x_fused: str | None = Header(default=None)):
    """Open a Claude Code session on this installation, about one failure.

    TWO WAYS IN, one endpoint:

      a failed row     `{job_id, title, detail, state, kind, message, page}` —
                       the download manager's own record of what broke.
      a description    `{note}` — the Preferences tab, where the user is
                       reporting behaviour rather than a crash. No traceback
                       exists, so the incident carries the app log instead and
                       the prompt tells the session to reproduce before it
                       diagnoses.

    One of `message`, `note` or `title` is required. Not as validation for its
    own sake: a session handed nothing at all has no failure, no description and
    no name for what it is looking at — it would read code at random and then
    write a report about having done so, which is worse than the button not
    working, because it costs the user minutes to find that out.

    Answers `{run_id, target, incident, report}`. The shell builds the sidebar
    URL from `target` + `run_id` (platform/lib/selffix.ts) rather than being
    handed one: view-vs-embed prefix is a fact about the document, and the
    server has no business guessing which one is asking.
    """
    guard = _require_fused(x_fused)
    if guard is not None:
        return guard
    if not isinstance(body, dict):
        return _error("request body must be a JSON object")

    # THE MACHINE BEFORE THE REQUEST. Whether Claude Code exists here is a fact
    # about the machine; whether the caller said what is wrong is a fact about
    # the request — and when both are unsatisfied the machine's answer is the
    # one that helps. Asking someone to describe a problem for a session that
    # cannot start on this computer is a form to fill in for nothing.
    #
    # It is also the ONLY answer the Preferences button can act on: "Set up
    # Claude Code" is offered with the description box empty (there is nothing
    # to describe yet — the CLI is the problem), so validating the body first
    # answered that click with "say what is wrong" and the user never saw the
    # install card the button exists to show.
    #
    # Same message and same status as the post-hoc path below, because it is the
    # same fact discovered a moment earlier: `spawn_helper` returns
    # CLAUDE_MISSING_ERROR when the CLI turns out to be missing, and that is
    # reported as 502.
    if _claude_cli_path() is None:
        return _error(claude_spawn.CLAUDE_MISSING_ERROR, status=502)

    # WHICH BRIEF the session gets, and it turns on whether the USER DESCRIBED
    # this — not on whether an error text came with it. A failed job row is
    # allowed to carry an empty `message` (jobs.py leaves it ""; the manager
    # renders a bare "Failed"), and keying off the message would hand that row
    # the Preferences brief: "NOTHING CRASHED, the user opened Preferences and
    # described something" — false on both counts, and it steers the session
    # away from tracing a failure that really happened. A `note` is the one
    # thing only the describe path produces.
    described = bool(str(body.get("note") or "").strip())
    reported_error = not described
    if not (described or str(body.get("message") or "").strip()
            or str(body.get("title") or "").strip()):
        return _error(
            "say what is wrong — a fix session needs either a failure to read "
            "or a description of what the app is doing wrong")

    root = selffix.install_root()
    # A READ-ONLY INSTALLATION STILL GETS A SESSION — a DIAGNOSTIC one (SF-4a).
    #
    # This used to be a 409, on the argument that a session which cannot write
    # spends minutes reading and then reports a fix that was never applied,
    # which to the user watching reads like a fix that was. That argument is
    # about a session that does not KNOW it cannot write. Told up front, the
    # same session is the most useful thing available on a machine nobody can
    # patch: it finds the cause and writes it down for someone who can, which is
    # a strictly better outcome than a refusal that leaves the user with the
    # original error and nothing else. An admin-installed copy is exactly where
    # that user is least able to help themselves.
    #
    # Everything downstream turns on this flag: the prompt (which must say so,
    # or the agent burns its turns fighting the permission error), where the
    # report goes (`records_dir` — outside a tree it could not write to), and
    # the watcher, which is skipped because nothing can change and therefore
    # nothing can be stamped.
    diagnostic = not selffix.writable()

    # Held across the LOOK and the SPAWN together, and released the moment the
    # request is done. Two clicks a millisecond apart would otherwise both find
    # the runs directory quiet and both start an agent; there is no state here
    # to leak, because what excludes the second session afterwards is the first
    # one's live process, not this lock.
    with _spawn_lock:
        busy = _live_session_in(root)
        if busy:
            return _error(
                f"a Claude session is already working on this installation "
                f"(run {busy}). Finish or stop it before starting another — two "
                "sessions editing the same files at once leave neither report "
                "true.", status=409)

        try:
            incident, report = selffix.record_incident(body)
        except OSError as exc:
            # Both records homes refused it. There is nowhere left to put the
            # file, and a session handed no incident has nothing to read.
            return _error(f"could not write the incident file: {exc}", status=500)

        # `writable()` PREDICTED; the write just PROVED. They disagree whenever
        # `os.access` says yes and something else says no — an ACL, a volume
        # remounted read-only under a running app, a full disk — and that used
        # to be a 500 on exactly the installation SF-13 exists to help. The
        # incident landing outside the tree is the proof, so believe it and run
        # the session in the mode that can actually finish.
        if not diagnostic and not selffix.in_state_dir(report):
            logger.info("self-fix: the install predicted writable and the write "
                        "disagreed — this session is diagnostic")
            diagnostic = True

        try:
            # BEFORE the session starts and never after. Two digests, not one:
            # the release's (for `reconcile`) and the tree as this session finds
            # it (what `settle` measures against) — see `selffix.begin_session`.
            # The incident write above cannot disturb either; the state dir is
            # outside the digest.
            #
            # SKIPPED WHEN DIAGNOSTIC, and not merely as an optimisation: the
            # baseline is written INTO the install tree, which is the thing this
            # branch cannot write to. There is also nothing for it to measure —
            # a digest exists to answer "did this session change the tree?", and
            # on a read-only install the answer is no by construction.
            before = "" if diagnostic else selffix.begin_session()[1]
        except OSError as exc:
            return _error(f"could not read the installation: {exc}", status=500)

        try:
            res = _spawn_helper(
                root,
                selffix.fix_prompt(incident, report, reported_error=reported_error,
                                   diagnostic=diagnostic),
                selffix.FIX_PERMISSION_MODE)
        except Exception as exc:  # noqa: BLE001 — the reason belongs in the response
            return _error(f"could not start the fix session: {exc}", status=502)
        run_id = res.get("run_id")
        if res.get("error") or not run_id:
            return _error(str(res.get("error") or "could not start the fix session"),
                          status=502)
        # Recorded INSIDE the lock and before it is dropped, so the next request
        # to take it already sees this run — the mutex only has to cover the
        # window in which nothing on disk names the session yet.
        selffix.note_session(str(run_id))

    # The marker's label for this fix. A described problem has no operation
    # name, so its first line stands in — the panel lists fixes by this, and
    # "a problem the user described" over every row says nothing.
    title = str(body.get("title") or "").strip()
    if not title:
        first_line = str(body.get("note") or "").strip().splitlines()
        title = first_line[0][:120] if first_line else ""
    # NOT WATCHED WHEN DIAGNOSTIC. The watcher exists to notice that the tree
    # changed and stamp the installation; on a read-only one it would poll a
    # digest that cannot move, and then try to write a marker into the same tree
    # it could not write the baseline to. The session is still perfectly real —
    # it just has nothing to stamp.
    if diagnostic:
        return {"run_id": str(run_id), "target": root, "incident": incident,
                "report": report, "diagnostic": True}
    try:
        threading.Thread(target=_watch_fix,
                         args=(str(run_id), incident, report, title, before),
                         daemon=True, name="fused-render-selffix-watch").start()
    except Exception:  # noqa: BLE001 — the session is already running
        # Nothing to unwind. The session is running and its run directory
        # already excludes a second one, whether or not anything is watching it
        # — which is the difference between a guard that is ASKED and one that
        # is remembered, and the reason this branch used to need an argument
        # about which harm to accept.
        #
        # One cost is left and it is not recoverable here: unwatched means
        # unstamped, so the badge will not appear for whatever this session
        # changes. The mark is a provenance claim only a watched session can
        # make (SF-7a) and is never inferred from a digest. Reaching this branch
        # at all means the interpreter could not start a thread.
        logger.exception(
            "could not start the self-fix watcher for run %s — the session is "
            "running unwatched and its changes will not be stamped", run_id)
    return {"run_id": str(run_id), "target": root, "incident": incident,
            "report": report}


@router.get("/api/selffix")
def api_selffix():
    """Everything the version chip's panel shows. See the module docstring for
    why this is not folded into /api/config."""
    snapshot = selffix.snapshot()
    # The registry's health rides along because this tab is where "something is
    # wrong with the app" lives (SF-14b), and because the toast that announces a
    # broken registry needs somewhere to send the user that holds the WHOLE
    # error and a way to copy it — a toast clamps at four lines.
    #
    # Added HERE rather than inside selffix.py, which deliberately imports
    # nothing from `server`. The router is the layer allowed to know both.
    error = _server_templates.registry_error()
    if error:
        snapshot["template_error"] = error
    return snapshot


@router.post("/api/selffix/clear")
def api_selffix_clear(x_fused: str | None = Header(default=None)):
    """Drop the modified marker without reinstalling.

    Offered because the badge is a claim about THIS machine that only the person
    at it can settle: they may have reverted the change by hand, or decided to
    keep it deliberately and not want a permanent banner about it. It clears the
    marker and nothing else — the report files stay on disk, so the record of
    what was changed survives the decision to stop being reminded about it.
    """
    guard = _require_fused(x_fused)
    if guard is not None:
        return guard
    return {"cleared": selffix.clear()}

File: frontend/src/platform/lib/selffix.ts
// Self-fix — asking Claude to repair this installation, and what it leaves
// behind (fused_render/selffix.py, SPEC §42).
//
// Two halves that meet at a file on disk:
//
//   THE TRIGGER    a failed row in the download manager grows one more option
//                  beside Dismiss. It POSTs the failure, gets back a live run,
//                  and lands the user in the explorer's chat sidebar OPEN ON
//                  THE INSTALLATION — so the fix is something they watch, not
//                  something that happens to their app.
//   THE MARK       if that session changed anything, the app stamps the
//                  install and the sidebar's version chip turns amber. Clicking
//                  it leads to the session's report, to the developers, and to
//                  a clean reinstall.
//
// The pure parts live here rather than inside the two components, because they
// are the parts that are wrong in ways a screenshot does not show: a deep link
// that drops the run id lands the user in an empty chat, and an issue URL that
// forgets the version is a bug report nobody can act on.
import { getJson, postJson } from "@platform/lib/api";
import { urlForFsPath } from "@platform/lib/router";

// One recorded fix — a session that changed this installation. Absolute paths:
// the server stores them relative to the install and resolves them on the way
// out, because an install can be moved.
export interface SelfFixEntry {
  at: number;
  updated_at: number;
  run_id: string;
  session_id: string;
  title: string;
  // The session's own account. Never null in practice — the server pre-creates
  // it holding the incident — but the marker predates nothing, so treat it as
  // optional rather than assume.
  report: string | null;
  incident: string | null;
}

// /api/config's `modified_install`: present ONLY when this installation has
// been changed by a fix session. Its absence is the unmodified state — there is
// no `modified: false` to check for.
export interface ModifiedInstall {
  modified: true;
  version: string | null;
  install_root: string;
  state_dir: string;
  first_modified_at: number | null;
  modified_at: number | null;
  fixes: SelfFixEntry[];
  // The report to open: the newest one. Chosen server-side so the chip and the
  // panel cannot disagree about which report "the" report is.
  latest_report: string | null;
}

export interface ReinstallAdvice {
  // brew | dmg | windows | linux | source | pip
  method: string;
  headline: string;
  // Empty when there is nothing to type — a DMG is dragged, not run. The panel
  // reads that as "the link IS the instruction" and promotes the link to the
  // section's primary action; see reinstall_advice's docstring.
  command: string;
  note: string;
  url: string;
  // Wording for the link, from the server (which already words the rest per
  // method). Older servers omit it — fall back to the bare URL.
  url_label?: string;
}

export interface SelfFixReport {
  path: string;
  name: string;
  at: number;
  size: number;
}

export interface SelfFixSnapshot {
  modified: boolean;
  version: string;
  install_root: string;
  // Whether a fix could be applied here at all. False for an installation the
  // user does not own — the panel says so instead of offering a fix that would
  // spend minutes and change nothing.
  writable: boolean;
  marker: ModifiedInstall | null;
  reports: SelfFixReport[];
  reinstall: ReinstallAdvice;
  issues_url: string;
  machine: Record<string, string | boolean>;
}

export interface SelfFixStart {
  run_id: string;
  target: string;
  incident: string;
  report: string;
}

// What the session is told is wrong — from EITHER of the two ways in.
//
//   a failed row   carries `message` (the error) and the row's own fields.
//   a description  carries `note`, and nothing else exists: no exception, no
//                  failed job. The server's incident then leans on the app log
//                  instead, and the prompt tells the session to reproduce
//                  before it diagnoses.
//
// Every field is optional on the wire, but the server requires at least one of
// `message` / `note` / `title`: a session handed nothing has nothing to look at.
export interface FailureContext {
  job_id?: string;
  title?: string;
  detail?: string;
  state?: string;
  kind?: string;
  message?: string;
  page?: string;
  /** The user's own words, when they are reporting behaviour rather than a
      crash (the Preferences tab). */
  note?: string;
  // Which surface offered the fix ("download manager"), so the report says how
  // the user got here.
  source?: string;
}

// -- The badge has to appear while you are watching ---------------------------
//
// The mark is set by the SERVER, mid-session, the moment the fix session's first
// edit lands (routers/selffix.py's watcher). Nothing pushes that to the shell —
// `modified_install` rides /api/config, which the sidebar reads once at boot.
// Left there, the badge for a fix you just watched happen would appear on your
// next launch, which is precisely when it is least intelligible.
//
// So the chip polls, at two cadences, the way the download manager does: slow
// forever, fast while a fix session plausibly runs. "Plausibly" is a timestamp
// this module writes when a fix STARTS — in localStorage rather than in memory,
// because the surface that starts the session (a download-manager row) and the
// surface that shows the badge (the sidebar chip) can be in different tabs, and
// because a reload during a long session must not drop back to the slow poll.
// THE STAMP: when a fix last started. Read to pick the cadence, and persisted
// rather than held in memory so a reload mid-session comes back in the fast lane.
export const SELFFIX_PING_KEY = "fused-render:selffix-ping";

// THE NUDGE: "self-fix state changed — re-read now", on two channels, because
// neither reaches everyone.
//
//   the KEY    fires `storage` in every same-origin document EXCEPT the writer,
//              so it reaches other tabs and never this one.
//   the EVENT  reaches this document and no other.
//
// Both are needed and neither is redundant, and getting that wrong is not a
// corner case in either direction: the surface that STARTS a fix (a
// download-manager row) and the one that DISMISSES a badge (the Preferences
// tab) are both normally in the very document the chip lives in, so the storage
// write alone would leave the badge stale for a full idle interval in exactly
// the cases the user is looking straight at it.
//
// ONE nudge for both events rather than two, because the listener's response is
// identical — cancel the pending timer, re-read now — and the cadence is decided
// separately, from the stamp above. A "started" nudge dispatched on a dismiss
// would be a lie; a second event with the same handler would be ceremony.
export const SELFFIX_CHANGED_KEY = "fused-render:selffix-changed";
export const SELFFIX_CHANGED_EVENT = "fused:selffix-changed";

// How long after a start the fast poll stays on. A session that takes longer
// than this has stopped being something the user is watching.
export const WATCH_WINDOW_MS = 30 * 60_000;
export const POLL_IDLE_MS = 60_000;
export const POLL_WATCH_MS = 5_000;

/** Whether a `storage` event is one of ours. Exported so the guard is pinned by
    a test rather than by a condition in a component: a guard that accepted only
    one of the two keys would silently swallow the other channel, which is a
    thing no screenshot shows. */
export function isSelfFixStorageKey(key: string | null): boolean {
  return key === SELFFIX_PING_KEY || key === SELFFIX_CHANGED_KEY;
}

/** "Something about self-fix changed — re-read." Both channels, always. */
export function notifySelfFixChanged(now = Date.now()): void {
  try {
    localStorage.setItem(SELFFIX_CHANGED_KEY, String(now));
  } catch {
    /* private mode / disabled storage — the idle poll is the floor */
  }
  try {
    window.dispatchEvent(new Event(SELFFIX_CHANGED_EVENT));
  } catch {
    /* no window (a test, a worker) — the idle poll is still the floor */
  }
}

export function noteFixStarted(now = Date.now()): void {
  try {
    // The stamp goes down FIRST: the nudge below makes listeners re-read the
    // cadence, and they must see this start rather than the previous one.
    localStorage.setItem(SELFFIX_PING_KEY, String(now));
  } catch {
    /* private mode / disabled storage — the idle poll is the floor */
  }
  notifySelfFixChanged(now);
}

export function lastFixStartedAt(): number {
  try {
    return Number(localStorage.getItem(SELFFIX_PING_KEY)) || 0;
  } catch {
    return 0;
  }
}

// A clock that has gone BACKWARDS past the stamp (a suspend, a manual clock
// change) reads as "not watching" rather than as watching forever: `now - at`
// negative fails the window test, which is the safe direction — a badge a
// minute late costs nothing, a permanent 5s poll costs a request every 5s for
// the life of the app.
export function selffixPollInterval(startedAt: number, now = Date.now()): number {
  const age = now - startedAt;
  return startedAt > 0 && age >= 0 && age < WATCH_WINDOW_MS
    ? POLL_WATCH_MS
    : POLL_IDLE_MS;
}

export async function startSelfFix(context: FailureContext): Promise<SelfFixStart> {
  const started = await postJson<SelfFixStart>("/api/selffix/start", context);
  // Only after the server said yes: a refused start (a read-only install) is
  // not a session anyone should be polling for.
  noteFixStarted();
  return started;
}

// A download-manager row as the fix session's brief. Kept here, next to the
// wire type, rather than inlined at the button: what a session is told about
// the failure is the difference between a useful diagnosis and a shrug, and the
// fields worth sending are not obvious from the row (the `page` attribution and
// the job id are what tie the report back to the call log).
//
// `Job` is deliberately not imported — that would tie the self-fix module to the
// job registry, when the trigger is meant to be usable from any surface that
// knows a title and a message. A structural parameter says the same thing and
// costs nothing.
export function failureContextFromJob(job: {
  id: string;
  title: string;
  detail: string;
  kind: string;
  state: string;
  message: string;
  page: string;
}): FailureContext {
  return {
    job_id: job.id,
    title: job.title,
    detail: job.detail,
    kind: job.kind,
    state: job.state,
    message: job.message,
    page: job.page,
    source: "download manager",
  };
}

// The Preferences tab's brief: the user's own description, and nothing else to
// go on. Trimmed here rather than at the button so "is there anything to send"
// is one answer — the Start control's enabled state and the request body must
// not disagree about whether a box holding three spaces counts.
export function failureContextFromNote(note: string): FailureContext {
  return { note: note.trim(), source: "preferences" };
}

export function describedProblemIsSendable(note: string): boolean {
  return note.trim().length > 0;
}

export function getSelfFix(): Promise<SelfFixSnapshot> {
  return getJson<SelfFixSnapshot>("/api/selffix");
}

// Dismissing is a state change like any other, and it can be made from EITHER
// surface — the chip's own popover or the Preferences tab. Nudging here rather
// than at each call site is what stops the two disagreeing: a dismiss from
// Preferences used to clear the marker and refresh that page while the sidebar
// chip stayed amber until its next idle poll, which reads as a dismiss that
// did not work.
export async function clearSelfFix(): Promise<{ cleared: boolean }> {
  const result = await postJson<{ cleared: boolean }>("/api/selffix/clear", {});
  notifySelfFixChanged();
  return result;
}

// Where the user lands once the session is running: the INSTALL FOLDER, with
// the chat sidebar open on the run that was just started.
//
// `_side=claude` is the explorer's companion-column param (listing/pane-side)
// and `run` is the chat template's re-attach param — the same pair the Inbox
// and scheduled messages hand over on (shell/schedule-lib's explorerUrl), and
// the same mechanism: neither param is in the pane's iframe src, the template
// reads them off the shell URL through fused.params' ancestor climb.
//
// The run id is what makes this a HANDOFF rather than a new chat. Without it
// the sidebar opens on the folder, sees no run, and shows an empty composer
// while a session is already working three feet away in a process nobody is
// watching.
export function fixSessionUrl(start: { target: string; run_id: string }): string {
  const params = new URLSearchParams({ _side: "claude", run: start.run_id });
  return urlForFsPath(start.target.replace(/[/\\]+$/, ""), "?" + params.toString());
}

// A GitHub issue with the machine facts already in it. The report itself is NOT
// inlined: it can be thousands of words, a URL that long is refused by servers
// long before it reaches GitHub, and the report is a file the user can attach
// or paste. What goes in the URL is the part they would otherwise have to be
// asked for twice — version, platform, and where the report is.
export function issueUrl(snapshot: SelfFixSnapshot): string {
  const version = snapshot.version || "unknown";
  const report = snapshot.marker?.latest_report || snapshot.reports[0]?.path || "";
  const body = [
    "<!-- Paste the self-fix report below, or attach the file. -->",
    "",
    "**Report file**: " + (report || "(none written)"),
    "",
    "| | |",
    "| --- | --- |",
    "| fused-render | v" + version + " |",
    "| platform | " + (snapshot.machine.platform ?? "unknown") + " |",
    "| python | " + (snapshot.machine.python ?? "unknown") + " |",
    "| install | `" + snapshot.install_root + "` |",
    "",
    "## What Claude changed here",
    "",
    "",
  ].join("\n");
  const params = new URLSearchParams({
    title: `Self-fix report — v${version}`,
    body,
    labels: "self-fix",
  });
  return `${snapshot.issues_url}?${params.toString()}`;
}

// The line under the chip's heading. Says WHEN, because "this install was
// modified" without a date is a fact the user cannot place — and the most
// common reaction to the badge is "when did that happen?".
export function modifiedSummary(marker: ModifiedInstall, now = Date.now()): string {
  const at = marker.modified_at;
  if (!at) return "Claude changed files in this installation.";
  const days = Math.floor((now / 1000 - at) / 86400);
  const when =
    days <= 0 ? "today" : days === 1 ? "yesterday" : `${days} days ago`;
  const count = marker.fixes.length;
  const what =
    count > 1 ? `${count} fix sessions have changed` : "A fix session changed";
  return `${what} files in this installation — most recently ${when}.`;
}
